netty/src/main/java/io/grpc/transport/netty/AbstractWriteQueue.java

```diff
@@ -0,0 +1,148 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A queue of pending writes to a {@link Channel} that is flushed as a single unit.
+ */
+public abstract class AbstractWriteQueue {
+
+  private static final int DEQUE_CHUNK_SIZE = 128;
+
+  /**
+   * {@link Runnable} used to schedule work onto the tail of the event loop.
+   */
+  private final Runnable later = new Runnable() {
+    @Override
+    public void run() {
+      channel.write(AbstractWriteQueue.this);
+    }
+  };
+
+  private final Channel channel;
+  private final BlockingQueue<Object> queue;
+  private final AtomicBoolean scheduled = new AtomicBoolean();
+
+  /**
+   * ArrayDeque to copy queue into when flushing in event loop.
+   */
+  private final ArrayDeque<Object> writeChunk = new ArrayDeque<Object>(DEQUE_CHUNK_SIZE);
+
+  public AbstractWriteQueue(Channel channel) {
+    this.channel = channel;
+    queue = new LinkedBlockingQueue<Object>();
+  }
+
+  /**
+   * Enqueue a flush on the channel.
+   */
+  void enqueueFlush() {
+    if (scheduled.compareAndSet(false, true)) {
+      // Add the queue to the tail of the event loop so writes will be executed immediately
+      // inside the event loop. Note DO NOT do channel.write outside the event loop as
+      // it will not wake up immediately without a flush.
+      channel.eventLoop().execute(later);
+    }
+  }
+
+  /**
+   * Enqueue a write command on the channel.
+   */
+  void enqueue(Object msg) {
+    queue.add(msg);
+    enqueueFlush();
+  }
+
+  /**
+   * Enqueue a write command with a completion listener.
+   */
+  void enqueue(Object msg, ChannelFutureListener completed) {
+    queue.add(new Object[]{msg, completed});
+    enqueueFlush();
+  }
+
+  /**
+   * Process the queue of commands and dispatch them to the stream. This method is only
+   * called in the event loop
+   */
+  void flush(ChannelHandlerContext ctx, ChannelPromise promise) {
+    try {
+      Object event;
+      boolean flushed = false;
+      while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
+        while (writeChunk.size() > 0) {
+          event = writeChunk.poll();
+          ChannelPromise eventPromise = ctx.newPromise();
+          if (event instanceof Object[]) {
+            eventPromise.addListener((ChannelFutureListener) ((Object[]) event)[1]);
+            event = ((Object[]) event)[0];
+          }
+          try {
+            handleCommand(event, eventPromise);
```

Comment:
This looks like it should be `channel.write(event, eventPromise)`. Then the Handlers don't get strange.